File: app/src/main/java/com/barryyang/barryyangdemo/designmodel/builder/AbstractCarModel.java
package com.barryyang.barryyangdemo.designmodel.builder;

import java.util.ArrayList;

/**
 * 车子模型
 *
 * @author barryyang
 */
public abstract class AbstractCarModel {

    private ArrayList<String> sequence = new ArrayList<>();

    /**
     * 车子启动
     */
    protected abstract void start();

    /**
     * 车子停下
     */
    protected abstract void stop();

    /**
     * 车子拧喇叭
     */
    protected abstract void alarm();

    /**
     * 引擎响声
     */
    protected abstract void engineBoom();

    final public void run() {
        for (String s : sequence) {
            if (s.equals("start")) {
                this.start();
            } else if (s.equals("stop")) {
                this.stop();
            } else if (s.equals("alarm")) {
                this.alarm();
            } else if (s.equals("engineBoom")) {
                this.engineBoom();
            }
        }
    }

    final public void setSequence(ArrayList<String> sequence) {
        this.sequence = sequence;
    }
}

File: app/src/main/java/com/barryyang/barryyangdemo/designmodel/model/AbstractCarModel.kt
package com.barryyang.barryyangdemo.designmodel.model

/**
 * @author : BarryYang
 * @date : 2020/8/24 9:51 AM
 * @desc :多个子类有公有的方法，并且逻辑基本相同时
 */
abstract class AbstractCarModel {

    abstract fun start()

    abstract fun stop()

    abstract fun engineBottom()

    abstract fun alarm()

    fun run() {
        this.start()
        this.engineBottom()
        this.alarm()
        this.stop()
    }

}